### DISCUSSION TITLE
Change Response

### DISCUSSION THREAD

**Discussion by hamedyosefian** [fastapi/fastapi#5762] [Questions] [answered]
Hi .. I have SQLALCHEMY And use this code for get all lessons 
```
@router_lesson.get('/lesson/', response_model=list[schemas.Lesson])
def read_lessons(db: Session = Depends(get_db)):
    response = crud.lesson.get_all(db)
    return response
```

Now i want add extra data each record and then return response...

The way I have now is that loop all record (response variable) and add data each record but not good way

What is the best way to do this?

**Comment by caeser1996** [comment]
You can modify the existing query to include the extra data as a calculated field using SQLAlchemy's query method. For example:
```
from sqlalchemy import func

@router_lesson.get('/lesson/', response_model=list[schemas.Lesson])
def read_lessons(db: Session = Depends(get_db)):
    response = (
        db.query(models.Lesson, func.count(models.UserLesson.id))
        .outerjoin(models.UserLesson)
        .group_by(models.Lesson)
        .all()
    )
    return [
        {
            "lesson": lesson,
            "extra_data": count,
        }
        for lesson, count in response
    ]

```
In this example, we are using the func.count method from SQLAlchemy to calculate the number of UserLesson records for each Lesson, and then grouping the results by the Lesson. The resulting query returns a list of tuples, where each tuple contains the Lesson record and the calculated count of UserLesson records.

You can then loop through the results and create a dictionary for each record that includes the Lesson record and the extra data, and return a list of these dictionaries.

This approach is more efficient than looping through the results after the query has been executed, as it reduces the number of database round trips.

**Reply by YuriiMotov** [reply]
This way response will not pass validation: response model is `list[schemas.Lesson]`, but you return `list[dict]` where every dict has fields `lesson` and `extra_data`.

**Comment by YuriiMotov** [comment, marked as answer]
Try this:

```python
from fastapi.applications import FastAPI
from pydantic import BaseModel


class LessonWithExtra(BaseModel):
    id: int
    name: str
    extra_field: str


app = FastAPI()

@app.get('/lesson/', response_model=list[LessonWithExtra])
def read_lessons():

    # response = crud.lesson.get_all(db)
    response = [
        {"id": 1, "name": "lesson 1"},
        {"id": 2, "name": "lesson 2"},
    ]

    return [dict(**lesson, extra_field="some value") for lesson in response]
```
